Updates to NFS and Volume Security topics

adellape opened the pull request:
Builds on https://github.com/openshift/openshift-docs/pull/1649.

Pretty build (internal):

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/install_config/persistent_storage/persistent_storage_nfs.html

adellape commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,68 @@
+[IMPORTANT]
 ====
+High-availability of storage in the infrastructure is left to the underlying
+storage provider.
```
@jeffvance I saw you removed this in your commit. I've added it back, but one section down. Did you feel strongly about dropping it?

jeffvance replied:
I dropped that sentence from the nfs doc and added it to the pod_security_context doc, since it is generic and applies to all of our supported storage.


----- Original Message -----
> >  ====
> > +High-availability of storage in the infrastructure is left to the
> > underlying
> > +storage provider.
> 
> @jeffvance I saw you removed this in your commit. I've added it back, but one
> section down. Did you feel strongly about dropping it?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files#r55769595

adellape commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,68 @@
+
+You must first create an object definition for the PV:
 
+.PV Object Definition Using NFS
```
@jeffvance I added this title back (more or less).

adellape commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +321,89 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
-- The firewall must be configured to allow traffic to the mount point, for NFSv4 the default port is 2049 (nfs).  
-  For NFSv3 there are three ports that need to be configured: 2049 (nfs), 20048 (mountd) and 111 (port mapper).
-+
-NFSv4:
```
I saw your commit dropped this NFSv3 vs NFSv4 iptables rules stuff, but I suspect it was only accidental (it was added pretty recently in general). Unless you had a particular reason for dropping it?

jeffvance replied:
As near as I can tell that must of been an accidental delete on my part, so good catch!

adellape commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +321,89 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
 # echo 'Y' > /sys/module/nfsd/parameters/nfs4_disable_idmapping
 ----
-
+|====
```
@jeffvance I turned this into a table? Not sold on it, trying it out.

jeffvance commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,68 @@
 toc::[]
 
 == Overview
-You can provision your OpenShift cluster with
```
Do we want to use words like "You" in this doc? I removed uses of "you", "your",  etc. thinking it was a bit more formal...

adellape replied:
@jeffvance Yeah, "you" / "your" is fine.

jeffvance replied:
Ok. I've left "your" and "you" in my GlusterFS and Ceph-RBD doc PRs (1729 & 1732)

jeffvance commented:
@adellape is there any way to move this pr along faster? We keep getting SA and customer requests that involve these docs.  Thanks.

pmorie commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,68 @@
+https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Storage_Administration_Guide/ch-nfs.html[NFS]
+is beneficial.
+
+The Openshift persistent volume (PV) framework allows administrators to provision a cluster
```
This doesn't sound quite right to me.  How about 'allows administrators to provision persistent storage for a cluster and gives...'

adellape replied:
+1

pmorie commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,68 @@
+The Openshift persistent volume (PV) framework allows administrators to provision a cluster
+with persistent storage and gives developers the ability to request those resources without
+having specific knowledge of the underlying infrastructure. Persistent volumes are not bound
+to a single project; they can be shared across the entire OpenShift cluster. Persistent volume
```
I would add a qualification to this sentence; as written it sounds to me like you can share PVs all over town when the reality is that claims cannot be shared across namespaces.  It might be better to introduce claims first, and say that claims are required to use a PV, and they are namespaced, but the volumes themselves can be claimed from any project.

pmorie replied:
Also, these paragraphs seem like they should be in a doc on PVs, not in the doc on NFS.  I would expect something like 'this doc explains how to use NFS with the PV framework. to learn more about the PV framework read this link'

adellape replied:
Agree, I don't love how we have this paragraph is basically duplicated across the 8 storage topics. See https://github.com/openshift/openshift-docs/pull/1726#discussion_r56526089.

pmorie commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +104,214 @@ pv0001                   <none>    5368709120   RWO           Available
+
+[NOTE]
+====
+PVs and PVCs are not necessary, just convenient, and make sharing a volume
```
I disagree with this sentence -- it is far more convenient to do less work, not more, when using a volume (my subjective opinion).  The real motivation for PV framework is to enable modeling the volumes themselves as cluster resources, to mediate access to those resources to avoid conflicts, and to be able to introspect who is using which storage.

adellape replied:
I've taken it out of a "NOTE" box and rewritten/reworked it into the Overview. New intro paragraphs:

> OpenShift clusters can be provisioned with persistent storage using NFS. Persistent volumes (PVs) and persistent volume claims (PVCs) provide a convenient method for sharing a volume across a project. While the NFS-specific information contained in a PV definition could also be defined directly in a pod definition, doing so does not create the volume as a distinct cluster resource, making the volume more susceptible to conflicts.

> This topic covers the specifics of using the NFS persistent storage type. Some familiarity with OpenShift and NFS is beneficial. See the Persistent Storage architecture topic for details on the OpenShift persistent volume (PV) framework in general.

jeffvance replied:
Also note that upstream there is an effort to disallow defining the storage plugin directly in the pod spec.
The proposal is to control, via a PSP "allow" setting, the pod's ability to define persistent storage inline.
The goal is to dissuade developers from doing this and direct them to use PVCs.
jeff

----- Original Message -----
> > +rather act as labels to match a PV to a PVC.
> > +<2> This claim will look for PVs offering *1Gi* or greater capacity.
> > +====
> > +
> > +Save the definition to a file, for example *_nfs-claim.yaml_*, and create
> > the
> > +PVC:
> > +
> > +====
> > +----
> > +# oc create -f nfs-claim.yaml
> > +----
> > +====
> > +
> > +[NOTE]
> > +====
> > +PVs and PVCs are not necessary, just convenient, and make sharing a volume
> 
> I've taken it out of a "NOTE" box and rewritten/reworked it into the
> Overview. New intro paragraphs:
> 
> > OpenShift clusters can be provisioned with persistent storage using NFS.
> > Persistent volumes (PVs) and persistent volume claims (PVCs) provide a
> > convenient method for sharing a volume across a project. While the
> > NFS-specific information contained in a PV definition could also be
> > defined directly in a pod definition, doing so does not create the volume
> > as a distinct cluster resource, making the volume more susceptible to
> > conflicts.
> 
> > This topic covers the specifics of using the NFS persistent storage type.
> > Some familiarity with OpenShift and NFS is beneficial. See the Persistent
> > Storage architecture topic for details on the OpenShift persistent volume
> > (PV) framework in general.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files/9a5c4affe7d76091de47c6b741c73c6702c25c43#r56526089

adellape replied:
Thanks for that context!

pmorie commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +321,89 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
 # setsebool -P virt_use_nfs 1
```
There's also a `virt_sandobx_use_nfs` that needs to be enabled.  This is one of the most frequent hangups people hit o this subject.

cc @swagiaal

jeffvance replied:
@pmorie I did *not* need to set `virt_sandbox_use_nfs on` in my testing. Are you certain? I'm wondering if there's something else going on in my test env...

adellape replied:
@pmorie @jeffvance `virt_sandbox_use_nfs` came up a few months back in https://github.com/openshift/openshift-docs/pull/1250#issuecomment-160999380. Do both need to be set here, and similarly both of the `fusefs` ones in the GlusterFS topic?

jeffvance replied:
Here are my nfs bools:
# getsebool -a|grep nfs|grep on\$
nfs_export_all_ro --> on
nfs_export_all_rw --> on
virt_use_nfs --> on

Note: the 2 export bools above can both be set off or mixed on/off with no effect on R/W access to my nfs export.

pmorie replied:
@swagiaal which of these do we need?  I'm positive people in the wild have needed both.

pmorie commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +321,89 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
-following:
+volume, each exported volume on the NFS server should conform to the following:
 
 - Each export must be:
```
This is no longer in the SELinux topic, new toc entry here?

jeffvance replied:
agreed.

adellape replied:
Fixed

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
+link:#fsgroup[`fsGroup`],
+link:#supplemental-groups[`supplementalGroups`],
+link:#user-id[`runAsUser`], and
+link:#selinux[`SELinuxOptions`].
```
case for SELinuxOptions should be `selinuxOptions` to match the other fields referred to here.

adellape replied:
`seLinuxOptions` (capital L) I think?

https://docs.openshift.com/enterprise/latest/rest_api/openshift_v1.html#v1-podsecuritycontext

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
+finally, ensuring that these ids are allowed in the range of legal ids defined for the
+project and/or the SCC that matches the requirements of the pod.
+
+Group ids, the user id, and selinux values are defined in the `*SecurityContext*` stanza
```
SELinux

pmorie replied:
@adellape Do you typically use the names of fields as they are in the golang type definitions, or as they surface to the user in API yaml/json?  If (2), this should be `selinuxOptions`

jeffvance replied:
In my opinion the field names we show in the docs should be what's in the yaml/json -- not in the code.

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
+image) will be validated against a range of allowable ids. If validation is required and
+fails then the pod will also fail.
+
+SCCs define strategies, such as `RunAsUser`, `supplementalGroups`, and `fsGroup`. These
```
`RunAsUser` and ( `supplementalGroups` and `fsGroup`) are not in the same category of things.

jeffvance replied:
They are all SCC stragegies, aren't they? What difference in these strategies is important to call out or re-categorize here?

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
+container (when those values are not supplied directly in the pod spec or image).
+
+SCCs may define the range of allowed ids (user, groups). If range checking is required
+(e.g., _MustRunAs_) and the allowable range is not defined in the SCC, then the project
```
Would restate as the value is validated against the range defined for the the project.

jeffvance replied:
The range might be defined in the SCC itself, and if so, that would supersede ranges defined in the project, correct?

pmorie replied:
The range is defined only on the project, as far as I know -- @pweil-?

jeffvance replied:
Paul, I've been able to leave the namespace alone and define ranges in the SCC, eg:
fsGroup:
  type: MustRunAs
  ranges:
  - min: 5000
    max: 6000
...
runAsUser:
  type: MustRunAsRange
  uidRangeMin: 99
  uidRangeMax: 65534
...
supplementalGroups:
  type: MustRunAs
  ranges:
  - min: 5000
    max: 6000


----- Original Message -----
> > +*_container_* defaults and not on Openshift policies. Here's a quick
> > summary of _RunAsAny_:
> > +
> > +- any id defined in the pod spec (or image) is allowed,
> > +- absence of an id in the pod spec (and in the image) results in the
> > container assigning an id,
> > +which is *root* (0) for docker,
> > +- no selinux labels are defined, so docker will assign a unique (and
> > generally not useful) label.
> > +
> > +For these reasons, SCCs with _RunAsAny_ for id related strategies should
> > be protected so
> > +that ordinary developers do not have access to the SCC.
> > +
> > +On the other hand, SCC strategies set to _MustRunAs_ or _MustRunAsRange_
> > trigger id validation
> > +(for id related strategies), and cause default values to be supplied by
> > Openshift to the
> > +container (when those values are not supplied directly in the pod spec or
> > image).
> > +
> > +SCCs may define the range of allowed ids (user, groups). If range checking
> > is required
> > +(e.g., _MustRunAs_) and the allowable range is not defined in the SCC,
> > then the project
> 
> The range is defined only on the project, as far as I know -- @pweil-?
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files/9a5c4affe7d76091de47c6b741c73c6702c25c43#r57159436

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
+if the SCC id strategy value is _MustRunAs_, and the minimum value of an id range is 100,
+and the id is absent from the pod spec, then 100 is provided as the default for this id.
+
+As part of pod authorization, the SCCs available to a pod are examined (roughly, in priority
```
s/authorization/admission/

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
````diff
@@ -11,144 +11,406 @@
+  namespace: default <3>
+...
+```
+<1> this is the name of the SCC that the pod used (in this case, a custom SCC)
````
s/this is// on this and the next two lines

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
````diff
@@ -11,144 +11,406 @@
+
+There are currently six predefined SCCs, seen below:
+```
+# oc get scc
````
@pweil- are there any new SCCs that should be added to this list?

In general I'm starting to feel like this section of docs should not be in a path relevant to persistent storage, this feels more like general security context and SCC docs to me.

pweil- replied:
Yep, hostnetwork.

jeffvance replied:
@pmorie perhaps it would be better to not list the predefined SCCs since this will likely change over time. Instead, just point out the SCCs relevant for the discussion below.

adellape replied:
> perhaps it would be better to not list the predefined SCCs since this will likely change over time. Instead, just point out the SCCs relevant for the discussion below.

+1

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
````diff
@@ -11,144 +11,406 @@
+Below is the _default_ namespace and a custom SCC, shown to summarize the interactions of
+the SCC and the namespace:
+```
+oc export ns default <1>
````
use `oc get ns default -o yaml` instead of `oc export`

jeffvance replied:
Sure. What is the reason for this? From what I recall the main difference is that `oc get...` has current status info.

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -11,144 +11,406 @@
-in the following example:
-
-====
+The _/opt/nfs/_ export is accessible by UID *99* and the group *5555*. In general, containers
```
I seem to remember that @jeffvance ran into major weirdness using uid 99 with busybox -- can we please pick another value?

jeffvance replied:
That's true and I forget what the wackiness was (maybe related to container's UID?). That is one reason for the couple of notes about UID 99 not being a particularly special id. `nfsnobody` and `nobody` seem like reasonable ids to use in these examples since that is not an unreasonable id to have on an export. Wrong?

pmorie replied:
@jeffvance 

The wackiness was that files owned by UID 99 were also owned by group 99, for no apparent reason.  I think we wound up figuring out that busybox treats 99 specially in certain ways, so I personally think you should find another UID, any other UID really, to use in this example, to avoid someone unwittingly hitting the same thing and filing issues / bugs around it.

pmorie commented on install_config/persistent_storage/pod_security_context.adoc:
````diff
@@ -11,144 +11,406 @@
+
+Here is a fragment of a new SCC named "nfs-scc":
+```
+# oc export scc nfs-scc 
````
`oc get scc nfs-scc -o yaml` is the better way to read an object like this, `oc export` blanks fields that cannot be set by the user.

jeffvance replied:
ah, so that's the main diff between `oc export` and `oc get -o yaml`...

jeffvance replied:
Thanks @pmorie !

adellape commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -11,67 +11,64 @@
-storage provider.
+_Project_ and _namespace_ are used interchangeably throughout this document. See
+link:../../architecture/core_concepts/projects_and_users.html#namespaces[Projects
+and Users] for details on the relationship.
```
@jeffvance Is there any reason I couldn't just update any references of "namespace" to "project" so we can lose this Note?

jeffvance replied:
Alex,

--all-namespaces is an oc command line option, so we expose namespace to users.
Also, "namespace" appears in some command output and in certain error messages.
My point being that we can't entirely hide the word "namespace", and that is
the reason I used both namespace and project in the docs, and why I have the note.
It's your call I just want you to know my motivation in using both.

jeff

----- Original Message -----
> >  ====
> > -High-availability of storage in the infrastructure is left to the
> > underlying
> > -storage provider.
> > +_Project_ and _namespace_ are used interchangeably throughout this
> > document. See
> > +link:../../architecture/core_concepts/projects_and_users.html#namespaces[Projects
> > +and Users] for details on the relationship.
> 
> @jeffvance Is there any reason I couldn't just update any references of
> "namespace" to "project" so we can lose this Note?
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files/9ce51e49014fad81a65119371df97a3505615842#r56526732

adellape replied:
@jeffvance Thanks. Yeah, I get the reasoning, and we don't hide "namespace" completely (https://docs.openshift.com/enterprise/latest/architecture/core_concepts/projects_and_users.html#namespaces).

I have just (in general, not just in this PR) wondered if it's worth liberally mixing the terms versus sticking with the OpenShift parlance (project, in this case) where we're able, to keep things simpler. We want readers generally to think in terms of projects, overall. I only saw one spot in these topics where "namespace" needed specifically to be used (a reference to the "namespace:" field, which could be explained in its callout description). Everywhere else, "project" looked like it would work / make sense just as well. Even the `oc export ns` looks like it could safely be `oc export project` (shows the same bits of info we're interested in here, AFAICT).

I think they're uncommon enough that for the spots where "namespace"-specific naming would need to be shown or used (e.g., "use --all-namespace to affect all projects"), we could just have an explanation or "Note" there as needed.

jeffvance replied:
Alex,

Good points and I'm fine changing all "namespace" to "project" and eliminating the note.
I will make that change to my recent gluster and ceph storage PRs.
jeff

----- Original Message -----
> >  ====
> > -High-availability of storage in the infrastructure is left to the
> > underlying
> > -storage provider.
> > +_Project_ and _namespace_ are used interchangeably throughout this
> > document. See
> > +link:../../architecture/core_concepts/projects_and_users.html#namespaces[Projects
> > +and Users] for details on the relationship.
> 
> @jeffvance Thanks. Yeah, I get the reasoning, and we don't hide "namespace"
> completely
> (https://docs.openshift.com/enterprise/latest/architecture/core_concepts/projects_and_users.html#namespaces).
> 
> I have just (in general, not just in this PR) wondered if it's worth
> liberally mixing the terms versus sticking with the OpenShift parlance
> (project, in this case) where we're able, to keep things simpler. We want
> readers generally to think in terms of projects, overall. I only saw one
> spot in these topics where "namespace" needed specifically to be used (a
> reference to the "namespace:" field, which could be explained in its callout
> description). Everywhere else, "project" looked like it would work / make
> sense just as well. Even the `oc export ns` looks like it could safely be
> `oc export project` (shows the same bits of info we're interested in here,
> AFAICT).
> 
> I think they're uncommon enough that for the spots where "namespace"-specific
> naming would need to be shown or used (e.g., "use --all-namespace to affect
> all projects"), we could just have an explanation or "Note" there as needed.
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files/9ce51e49014fad81a65119371df97a3505615842#r56541115

adellape commented:
> is there any way to move this pr along faster? We keep getting SA and customer requests that involve these docs. Thanks.

@jeffvance @pmorie Sorry, was PTO yesterday. Still WIP (i.e., most of the `pod_security_context.adoc` still to go), but I've made some edits based on latest comments and updated the pretty build:

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/install_config/persistent_storage/persistent_storage_nfs.html

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/architecture/additional_concepts/storage.html

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/install_config/persistent_storage/pod_security_context.html

Continuing to work on this today.

adellape commented:
@jeffvance @pmorie Finished going through them, PTAL:

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/install_config/persistent_storage/persistent_storage_nfs.html

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/architecture/additional_concepts/storage.html

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/install_config/persistent_storage/pod_security_context.html

I addressed most of the comments so far, save for the following threads that still have some outstanding questions:

https://github.com/openshift/openshift-docs/pull/1726#discussion_r56565553
https://github.com/openshift/openshift-docs/pull/1726#discussion_r56416651
https://github.com/openshift/openshift-docs/pull/1726#discussion_r56416899
https://github.com/openshift/openshift-docs/pull/1726#discussion_r56418155

jeffvance commented on architecture/additional_concepts/storage.adoc:
```diff
@@ -13,15 +13,23 @@ toc::[]
+and are created and used by developers. A PVC is required if a developer wants
+to use a PV. The PVs themselves, however, are not bound to a single project;
+they can be shared across the entire OpenShift cluster and claimed from any
+project.
```
after this sentence:
 "...they can be shared across the entire OpenShift cluster and claimed from any project." 
add:
Note that once a PV has been bound to a PVC that PV can not be bound to additional PVCs. This has the effect of scoping a bound PV to a single namespace -- that of the binding project.

adellape replied:
Updated to the following:

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/architecture/additional_concepts/storage.html#overview

jeffvance commented on architecture/additional_concepts/storage.adoc:
```diff
@@ -32,12 +40,12 @@ High-availability of storage in the infrastructure is left to the underlying
+node resources and PVCs consume PV resources. For example, pods can request
+specific levels of resources (e.g., CPU and memory), while PVCs can request
+specific link:#pv-capacity[storage capacity] and link:#pv-access-modes[access
+modes] (e.g, they can be mounted once read/write or many times read-only).
```
@pmorie is the sentence above correct -- "e.g, they can be mounted once read/write or many times read-only"?
My understanding is that `accessModes` has nothing at all to do with actual access. Instead `accessModes` are merely labels to help match a PV and a PVC. Assuming that is true, then do these labels really impact how a volume is mounted?

jeffvance commented:
@adellape I will re-test nfsnobody to address the last of your 4 comment questions above. @pmorie needs to reply to the first 3 of those comment questions.

tnguyen-rh commented on architecture/additional_concepts/storage.adoc:
```diff
@@ -13,15 +13,26 @@ toc::[]
+PVCs are specific to a
+link:../../architecture/core_concepts/projects_and_users.html#project[project]
+and are created and used by developers. A PVC is required if a developer wants
+to use a PV. The PVs themselves are not bound to a single project; they can be
```
@adellape 
Perhaps the first two sentences could be combined to make it flow better, something like: "PVCs are specific ...., and are created and used by developers as a means to use a PV."
Also, if "bind" (or "bound", here) is a technical term, it should be italicized on first use.  (Maybe it is, above, and i don't see it, in which case feel free to ignore this portion of the comment.)

adellape replied:
+1. Now it's:

> PVCs are specific to a project and are created and used by developers as a means to use a PV. PV resources on their own are not scoped to any single project; they can be shared across the entire OpenShift cluster and claimed from any project. After a PV has been bound to a PVC, however, that PV cannot then be bound to additional PVCs. This has the effect of scoping a bound PV to a single namespace (that of the binding project).

"bound" is italicized and links to `#binding`.

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +93,207 @@ pv0001                   <none>    5368709120   RWO           Available
-can be its own export. Each export is one persistent volume. Kubernetes enforces
-unique names for persistent volumes, but the uniqueness of the NFS volume's
-server and path is up to the administrator.
+Disk partitions can be used to enforce disk quotas and size constraints. Each
```
@adellape 
Passive voice.  Suggest: "OpensShift enforces d.q. & s.c. via disk partitions".

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +93,207 @@ pv0001                   <none>    5368709120   RWO           Available
-processes handle reclamation tasks based on policies set on each persistent
-volume.
+The *_/etc/exports_* file on the NFS server contains the accessible NFS
+directories. The target NFS directory has the familiar POSIX owner and group
```
Maybe s/familiar// ?

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +93,207 @@ pv0001                   <none>    5368709120   RWO           Available
+----
+<1> `*securityContext*` must be defined globally to the pod, not under a
+specific container.
+<2> An array of GIDs defined globally for the pod.
```
Suggest being explicit for callout 2, adding something along the lines of "in this case, there is one element in the array -- more GIDs would be comma-separated", for the benefit of readers unfamiliar w/ the inline array YAML syntax.

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +93,207 @@ pv0001                   <none>    5368709120   RWO           Available
+[[nfs-user-ids]]
+=== User IDs
+
+User IDs can be defined in the container image or in the pod definition.
```
I'll register a general gripe w/ "can be" constructs, here...

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -99,66 +93,207 @@ pv0001                   <none>    5368709120   RWO           Available
+- 99 is not included in the SCC or project's user ID range.
+
+It is generally considered a good practice not to modify the predefined SCCs.
+The preferred way to fix this situation is to create a custom SCC, as described
```
The "fix this situation" implies brokenness, which isn't the intent (i believe).  Suggest: "It is generally considered a good practice to leave the predefined SCCs unmodified, and create a custom SCC..." although that doesn't sound so great, either....

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +303,92 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
+
+The *virt_use_nfs* boolean is defined by the *_docker-selinux_* package. If an
+error is seen indicating that this bool is not defined, then ensure that this
+package has been installed.
```
How about "ensure this package is installed"?

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +303,92 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
-volume, each exported volume on the NFS server itself should conform to the
-following:
+In order to enable arbitrary container users to read and write the volume, each
+exported volume on the NFS server should conform to the following:
```
Suggest: "... following conditions." (w/ full stop).  This will make it easier to translate, i believe.

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +303,92 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
-NFSv4:
+- The firewall must be configured to allow traffic to the mount point. For NFSv4,
+the default port is 2049 (*nfs*). For NFSv3, there are three ports that need to
+be configured: 2049 (*nfs*), 20048 (*mountd*), and 111 (*portmapper*).
```
Suggest: "three ports to configure: ..."

tnguyen-rh commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -168,56 +303,92 @@ To enable writing to NFS volumes with SELinux enforcing on each node, run:
-[[nfs-additional-config-and-troubleshooting]]
+[[nfs-automation]]
+== Automation
+As discussed, clusters can be provisioned with persistent storage using NFS in
```
Maybe add an xref to where it was discussed?

adellape replied:
The links in the bulleted items do that, but I've just dropped the "As discussed" lead-in.

ahardin-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+also fail.
+
+SCCs define strategies, such as `*runAsUser*`, `*supplementalGroups*`, and
+`*fsGroup*`. These strategies are used to help decide if the pod is authorized.
```
s/are used to help decide/help decide

ahardin-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+For these reasons, SCCs with *RunAsAny* for ID-related strategies should be
+protected so that ordinary developers do not have access to the SCC. On the
+other hand, SCC strategies set to *MustRunAs* or *MustRunAsRange* trigger ID
+validation (for ID related strategies), and cause default values to be supplied
```
s/ID related strategies/ID-related strategies  (as seen in the previous sentence)

tnguyen-rh replied:
Missing hyphen: "ID-related"

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+- understanding SELinux considerations, and
+- ensuring that these IDs are allowed in the range of legal IDs defined for the
+project and/or the SCC that matches the requirements of the pod.
+
```
(again, for translation) Better to restructure as a complete sentence as intro, and complete sentences as the points.

ahardin-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+support ranges of allowable ID. However, unlike SCCs, projects do not define
+strategies, such as `*runAsUser*`.
+
+Allowable ranges are interesting not only because they define the boundaries for
```
Not sure I like the use of "interesting" here. Do you mean "important" or "helpful"? You may want to rephrase for clarity's sake.

adellape replied:
Went with "helpful".

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+- Absence of an ID in the pod definition (and in the image) results in the
+container assigning an ID, which is *root* (0) for Docker.
+- No SELinux labels are defined, so Docker will assign a unique (and generally not
+useful) label.
```
Suggest omitting "(and generally not useful)".

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+container IDs, but also because the minimum value in the range becomes the
+default value for the ID in question. For example, if the SCC ID strategy value
+is *MustRunAs*, and the minimum value of an ID range is 100, and the ID is
+absent from the pod definition, then 100 is provided as the default for this ID.
```
Maybe too many ", and CONDITION" clauses?  Suggest removing first "and".

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+Any SCC with a strategy set to *RunAsAny* allows arbitrary values for that
+strategy to be defined in the pod definition (and/or image). When this applies
+to the user ID (`*runAsUser*`) it is prudent to restrict access to the SCC to
+prevent a container from being able to run as root.
```
Nice: "prudent".  :smile:

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+Because pods often match the *restricted* SCC, it is worth knowing the security
+this entails. The following observations apply to the *restricted* SCC:
+
+- Constrains user IDs due to the `*runAsUser*` strategy being set to
```
I like "observations apply", personally, but it probably sounds better in an academic journal than here.  Suggest: "The *restricted* SCC has several characteristics", followed by complete sentences in the points.

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+default group ID is the starting ID in the first range, which is `1000000000` in
+this project. If the SCC did not define a minimum group ID, then the project's
+default ID is applied.
+<5> Same as *<4>* but for user IDs. Also, only a single range of user IDs is
```
These callouts are humongous.  Perhaps the common syntax (M/N, M-N) explanation could be put into  a small NOTE or somesuch.

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+this topic apply to any non-root owner.
+====
+
+The *_/opt/nfs/_* export is accessible by UID *99* and the group *5555*. In
```
ISTR directories should end in slash, but then there are several instances in the docs of them w/o.  (shrug)

adellape replied:
:(

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+- the group ID of 5555 is allowed.
+
+It is considered a better practice to create new SCCs versus modifying a
+predefined SCC, or changing the range of allowed IDs in the predefined projects.
```
Hmm this sounds familiar.  However, "better practice" is weird to my ears.

adellape replied:
s/considered a better practice/better/

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+files created in the volume.
+====
+
+As is true with `*supplementalGroups*`, all containers in the above pod
```
s/is true //

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+- ID range checking is enforced,  and
+- the group ID of 5555 is allowed.
+
+It is considered a better practice to create new SCCs versus modifying a
```
:/

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+has access to the new SCC), it will start because the group *5555*, supplied in
+the pod definition, is allowed by the custom SCC. Additionally, the pod will
+"take over" the block device, so when the block storage is viewed by a process
+outside of the pod, it will actually have 5555 as its group permissions.
```
I'm confused: 5555 is the group id or is it a permissions (mode bits) mask?

ahardin-rh replied:
s/5555/*5555*

adellape replied:
s/permissions/ID/

ahardin-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+critical. Sometimes, the SCC does not define a range of allowable group IDs but
+requires group ID validation (due to `*supplementalGroups*` set to *MustRunAs*;
+note this is not the case for the *restricted* SCC). The project will not likely
+allow a group ID of 5555, unless the project has been customized for access to
```
s/5555/*5555*  (to be consistent with the rest of the instances in the paragraph)

ahardin-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
-groups.
+- a minimum and max group ID are defined,
+- ID range checking is enforced, and
+- the group ID of 5555 is allowed.
```
s/5555/*5555*  (to be consistent with the above content)

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+Similar to group IDs, user IDs may be validated according to policies set in the
+SCC and/or project. If the SCC's `*runAsUser*` strategy is set to *RunAsAny*,
+then any user ID defined in the pod definition or in the image is allowed. This
+means even a UID of 0 (root) is allowed!
```
Too bad we can't include a "freaking out" emoticon...

ahardin-rh replied:
As exciting as this statement is, I suggest removing the exclamation   :grinning:

adellape replied:
@tnguyen-rh :goberserk:

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+supplies no user ID, then the default ID is the minimum value of the range of
+allowable user IDs.
+
+Getting back to the earlier link:#nfs-example[NFS example], the container needs
```
s/Getting back/Returning/

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+
+To fix this situation:
+
+- A new SCC could be created with the appropriate user ID range (recommended).
```
Suggest converting "(recommended)" into a separate sentence.

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+
+- A new SCC could be created with the appropriate user ID range (recommended).
+- The *restricted* SCC could be modified to include 99 within its minimum and
+maximum user ID range (not recommended).
```
idem
Bonus would be to explain why (both cases).

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+- The *default* project's UID range could be changed to allow a user ID of 99 (not
+generally advisable because only a single range of user IDs can be specified).
+- A new project could be created with the appropriate user ID range defined (not
+covered here).
```
Generally, the parenthesized blurbs should either be eliminated (in the case of "not covered here") or converted to more substantial separate sentences.

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+[[scc-runasuser]]
+*User IDs and Custom SCCs*
+
+It is generally considered a good practice to not modify the predefined SCCs.
```
hmm, "best", "better", "good" -- i see a pattern here...

tnguyen-rh commented on install_config/persistent_storage/pod_security_context.adoc:
```diff
@@ -12,262 +12,728 @@ toc::[]
+The SCC determines whether or not to require an SELinux label and can provide a
+default label. If the `*seLinuxContext*` strategy is set to *MustRunAs* and the
+pod (or image) does not define a label, a default is used either from the SCC
+itself or from the project.
```
Suggest: "... label, OpenShift defaults to a label chosen from ..."

tnguyen-rh commented:
@adellape 
Lots of text, but a normal nit/suggestion density.

ahardin-rh commented:
@adellape just a few minor nits from me

adellape commented:
> I think you should probably s/99/65534/g (which is defined as "nfsnobody).

https://github.com/openshift/openshift-docs/pull/1726/commits/e2a53d732c95646e2a5cab715ca86f9edae5019f

@jeffvance ^ WDYT? (I'll squash this down at the end.) I left the 5555 example group as is.

adellape commented:
@tnguyen-rh @ahardin-rh Thx for the review. Pretty much all comments are address, even if I didn't comment/reply as such inline. Pretty build updated too. I'll squash my commits before merging.

@jeffvance Removing WIP from this. Will wait for @pmorie to address the following outstanding items before merging, though:

https://github.com/openshift/openshift-docs/pull/1726#issuecomment-198095836
https://github.com/openshift/openshift-docs/pull/1726#issuecomment-198435520

jeffvance commented on architecture/additional_concepts/storage.adoc:
```diff
@@ -20,11 +20,11 @@ specific knowledge of the underlying storage infrastructure.
+entire OpenShift cluster and claimed from any project. After a PV has been
+link:#binding[_bound_] to a PVC, however, that PV cannot then be bound to
+additional PVCs. This has the effect of scoping a bound PV to a single
 link:../../architecture/core_concepts/projects_and_users.html#namespaces[namespace]
```
Line 155: ifdef::openshift-enterprise,openshift-origin[]
In my view of this doc I see the following:
---------------------------------------------
   Types of Persistent Volumes
   OpenShift supports the following PersistentVolume plug-ins:

   Capacity
---------------------------------------------
In other words, the volume plug-ins are omitted. I think the ifdef should appear before "Types of Persistent Volumes"

adellape replied:
Are you looking at the GitHub render of the doc? It won't respect ifdefs. It should be visible in the built versions of the docs:

http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/architecture/additional_concepts/storage.html#types-of-persistent-volumes

jeffvance replied:
Hi Alex,  I was using the github rendering (I forgot). My mistake...


----- Original Message -----
> > @@ -20,11 +20,11 @@ specific knowledge of the underlying storage
> > infrastructure.
> >  
> >  PVCs are specific to a
> >  link:../../architecture/core_concepts/projects_and_users.html#project[project]
> > -and are created and used by developers. A PVC is required if a developer
> > wants
> > -to use a PV. The PVs themselves are not bound to a single project; they
> > can be
> > -shared across the entire OpenShift cluster and claimed from any project.
> > After a
> > -PV has been bound to a PVC, however, that PV cannot be bound to additional
> > PVCs.
> > -This has the effect of scoping a bound PV to a single
> > +and are created and used by developers as a means to use a PV. PV
> > resources on
> > +their own are not scoped to any single project; they can be shared across
> > the
> > +entire OpenShift cluster and claimed from any project. After a PV has been
> > +link:#binding[_bound_] to a PVC, however, that PV cannot then be bound to
> > +additional PVCs. This has the effect of scoping a bound PV to a single
> >  link:../../architecture/core_concepts/projects_and_users.html#namespaces[namespace]
> 
> Are you looking at the GitHub render of the doc? It won't respect ifdefs. It
> should be visible in the built versions of the docs:
> 
> http://file.rdu.redhat.com/~adellape/031016/storage-sec-update/architecture/additional_concepts/storage.html#types-of-persistent-volumes
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/openshift-docs/pull/1726/files/e2a53d732c95646e2a5cab715ca86f9edae5019f..2a8f749d67480f3056c4973e786bc4204d8ef3d2#r57059097

jeffvance commented on install_config/persistent_storage/persistent_storage_nfs.adoc:
```diff
@@ -356,15 +357,14 @@ claim.
+- Enforce security by link:#nfs-volume-security[restricting volumes] to the
+project that has a claim to them.
+- Configure link:#nfs-reclaiming-resources[reclamation of discarded resources] for
+each PV.
```
Under "Group IDs", in the pod fragment example, `runAsUser` is comment out and with a comment stating "65534 from above has been commented out here". At some point in my changes I removed this line since I moved group ID before user ids. (Previously, when user ids came before group ids this comment made more sense.) I think the `runAsUser` stanza should be removed.

adellape replied:
OK, I've removed this line from the Group IDs example:

```
      #runAsUser: 65534 from above has been commented out here
```

adellape commented:
@jeffvance @pmorie @erinboyd It looks like there's a few outstanding questions still, but for now I've added `virt_sandbox_use_nfs` as a bool (in addition to `virt_use_nfs`), squashed, and will merge this to move things along. We can follow-up if needed.

adellape commented:
[rev_history]
|link:../architecture/additional_concepts/storage.html[Additional Concepts -> Persistent Storage]
|Updated the link:../ink:../architecture/additional_concepts/storage.html[Overview] section to provide a better introduction to the PV and PVC concepts.
%
.3+|link:../install_config/persistent_storage/persistent_storage_nfs.html[Configuring Persistent Storage -> Persistent Storage Using NFS]
|Updated for clarity throughout.

|Enhanced the link:../install_config/persistent_storage/persistent_storage_nfs.html#nfs-volume-security[Volume Security] section significantly.

|Added the link:../install_config/persistent_storage/persistent_storage_nfs.html#nfs-additional-config-and-troubleshooting[Additional Configuration and Troubleshooting] section.
%
|link:../install_config/persistent_storage/pod_security_context.html[Volume Security]
|Updated significantly for clarity throughout.
%

adellape commented:
Just realized this should have been marked for cherrypick to the dedicated-3.1 and online branches, because of the two Arch topics that were updated (however minor). Added the labels, I'll see about doing the picking them myself async, or I might throw it back into the Next Release milestone for inclusion next week (it's already been picked for enterprise-3.1 but the script should still work for dedicated/online).

adellape commented:
Yeah I think I'm gonna just move it back to Next Release for next week, since it needs to get picked at the same time with https://github.com/openshift/openshift-docs/pull/1833 (which is currently in Next Release).

vikram-redhat commented:
@adellape if this is already in the enterprise 3.1 branch, should the enterprise 3.1 label be removed from it?

adellape commented:
@vikram-redhat My thought was to leave them for posterity's sake. If they're picked again to the other branches, it should go in as a "no change" pick and not disrupt anything.
